image/lib.sh

```diff
@@ -25,23 +32,68 @@ function errnullify     { "$@" 2>/dev/null; }
 : ${RUNTIME_DEFERRED:="$IMAGE_ROOT/runtime.d"}
 
 ##
-## Utils
+## Utils (must be in-process, hence placement here).
+##
+
+abspath() {
+	(cd "$1" && pwd)
+}
+
+first-in() {
+	local predicate="$1"
+	shift 1
+
+	local item
+	for item in "$@"; do
+		$predicate "$item" || continue
+
+		echo "$item"
+		return
+	done
+
+	return 1
+}
+
+abself() {
+	local dir base
+	dir=$(abspath "${PWD%/}/${0%/*}")
+	base="${0##*/}"
+	echo "$dir/$base"
+}
+
+save-retval() {
+	local -n rv_ref=$1; shift
```

Comment:
Bash now supports real indirect referencing, which is really quite nice to have for things like this.